### PR TITLE
Invoke tsc programmatically

### PR DESCRIPTION
**Pull request by chriskrycho**

### CODE REVIEW

**Comment by dzaman** [comment]
**TL;DR:** After installing `ember-cli-typescript` on an existing app I had trouble with automatic rebuilds / live reloading. Switching to the `programmatic-compilation` build solved this problem.

---

The initial build would work as expected in a normal amount of time for our app (20s), but rebuilds for me would take significantly longer (in some cases 5+ minutes) and in other cases I gave up waiting. I think it's possible that the completed builds were actually tied to additional file changes and would have otherwise hung indefinitely. 

When the rebuilds hung, I'd see output that a file has changed and that incremental compilation was starting, but neither the build success message nor the expected typescript warnings were output.

I would kill rebuilds older than that, but wasn't able to `ctrl+c` them, I'd have to `kill` the process. I'd see the "cleaning up..." message, but it would just hang after that. 

I'm using  `Ember 2.16.2` on `OS X Sierra 10.12.6`

**Comment by jasonworden** [comment]
While trying @dzaman's branch with `ember-cli-typescript` on that same app, I had similar troubles. Here is what happened.
- Initial build took expected amount of time
- I was able to make 5-10 changes throughout a ~15 minute span that were reflected in the browser via hot-reloaded as expected.
- I made a subsequent change, and my browser tab for the app began to refresh but was hung up for several minutes. The re-build after these changes never finished.
- I ctrl-C'd out of the `ember server` process, but then had to `kill` the process.
- Chrome crashed, I believe right after I killed the server process.
- Restarted the server and was able to make a few changes with hot-reloaded working for about 5-10 minutes.
- A subsequent change did not result in a proper re-build, and I followed the same steps.
- I changed app branches and was able to work without the Typescript in the project.
- I'm working with the `programmatic-compilation` now and everything seems good!

`Ember 2.16.2` on `OS X Sierra 10.13.1`

**Comment by jasonworden** [comment]
While trying @dzaman's branch with `ember-cli-typescript` on that same app, I had similar troubles. Here is what happened.
- Initial build took expected amount of time
- I was able to make 5-10 changes throughout a ~15 minute span that were reflected in the browser via hot-reloaded as expected.
- I made a subsequent change, and the browser tab did not reflect the changes. I manually refreshed the browser tab, and it was hung up trying to refresh for several minutes. The re-build after these changes never finished.
- I ctrl-C'd out of the `ember server` process, but then had to `kill` the process.
- Chrome crashed, I believe right after I killed the server process.
- Restarted the server and was able to make a few changes with hot-reloaded working for about 5-10 minutes.
- A subsequent change did not result in a proper re-build, and I followed the same steps.
- I changed app branches and was able to work without the Typescript in the project.
- I'm working with the `programmatic-compilation` now. Everything seemed good for awhile, but I eventually ran into the same issue as above.

`Ember 2.16.2` on `OS X Sierra 10.13.1`

**Comment by chriskrycho** [comment]
@dfreeman one thing I was thinking about just now: can we leverage this to properly fail builds when people have the `noEmitOnError` setting enabled? That'd be a nice win here.

**Comment by machty** [comment]
This PR fixes an [issue with ember-cli-addon-docs](https://github.com/ember-learn/ember-cli-addon-docs/issues/144) that I was running into where changing a `.md` file caused the rebuild to hang. I believe the root cause is that the `tsc` binary doesn't emit the "Compilation complete. Watching for file changes." text when it sees a non-JS/TS file change.

**Comment by tansongyang** [comment]
Hello everyone. I'm checking in here to see what else needs to be done before this can be merged into `master`.

**Comment by chriskrycho** [comment]
We need to get the tests passing, either by removing the helper-using test or by fixing the typings so it works properly (preferably but not *necessarily* the latter).

**Comment by tansongyang** [comment]
@chriskrycho Is work already being done on the typings? If not, I think I could give some time to that this weekend.

**Comment by chriskrycho** [comment]
Not to my knowledge. I haven’t had time to actually figure out what’s wrong there.

**Comment by tansongyang** [comment]
OK. If I don't hear any updates by this weekend, I'll just start on it. Could you point me to where I should be looking?

**Comment by chriskrycho** [comment]
@tansongyang what we know so far is documented in #133.

**Comment by chriskrycho** [comment]
That is one super weird AppVeyor failure. Do we have some kind of caching problem, maybe?

**Comment by dfreeman** [comment]
@chriskrycho Haha, I was just about to ping you to ask if we could try to clear that somehow. Bad cached dependency tree, maybe? 😬

**Comment by jamescdavis** [comment]
:tada:
